sage_modabvar/abvar.py

```diff
@@ -2196,20 +2201,56 @@ def frobenius_polynomial(self, p):
             sage: J=J0(23)
             sage: J.frobenius_polynomial(997)
             x^4 + 20*x^3 + 1374*x^2 + 19940*x + 994009
+
+            sage: J = J0(33)
+            sage: J.frobenius_polynomial(7)
+            x^6 + 9*x^4 - 16*x^3 + 63*x^2 + 343
+
+            sage: from sage_modabvar import J1
+            sage: A = J1(27)[1]; A
+            Simple abelian subvariety 27bG1(1,27) of dimension 12 of J1(27)
+            sage: A.frobenius_polynomial(11)
+            x^24 - 3*x^23 - 15*x^22 + 126*x^21 - 201*x^20 - 1488*x^19 + 7145*x^18 - 1530*x^17 - 61974*x^16 + 202716*x^15 - 19692*x^14 - 1304451*x^13 + 4526883*x^12 - 14348961*x^11 - 2382732*x^10 + 269814996*x^9 - 907361334*x^8 - 246408030*x^7 + 12657803345*x^6 - 28996910448*x^5 - 43086135081*x^4 + 297101409066*x^3 - 389061369015*x^2 - 855935011833*x + 3138428376721
+
         """
+        if not is_prime(p):
+            raise ValueError("p must be prime")
+        if not self.is_simple():
+            return prod((s.frobenius_polynomial(p) for s in
+                         self.decomposition()))
+
         f = self.newform('a')
         Kf = f.base_ring()
-        name = Kf._names[0]
-        ap = f.modular_symbols(1).eigenvalue(p, name)
         eps = f.character()
-        L = Kf.galois_closure('a')
-        emb = Kf.embeddings(L)[0]
+        Qe = f.base_ring()
+        d = Kf.degree() / Qe.degree()
 
-        R = PolynomialRing(ZZ, 'x')
+        R = PolynomialRing(QQ, 'x')
+        S = PolynomialRing(ZZ, 'x')
         x = R.gen()
-        return R(prod([x**2 - sigma(emb(ap))*x + sigma(emb(eps(p)))*p 
-            for sigma in L.embeddings(L)]))
 
+        if Kf != QQ:
+            # relativize number fields to compute charpoly of
+            # left multiplication of ap on Kf as a Qe-vector
+            # space.
+            Lf = Kf.relativize(Qe.gen(), 'a')
+            to_Lf = Lf.structure()[1]
+
+            name = Kf._names[0]
+            ap = to_Lf(f.modular_symbols(1).eigenvalue(p, name))
+
+            Gp = ap.matrix().charpoly()
+            points = [(Y, Qe(Y**d * Gp(Y + to_Lf((eps(p)) * p) / Y)).norm())
+                      for Y in range(1, 2*Kf.degree()+2)]
+        else:
+            ap = f.modular_symbols(1).eigenvalue(p)
+            Gp = x - ap
+
+            points = [(Y, Qe(Y**d * Gp(Y + (eps(p) * p) / Y)).norm())
```

Comment:
If you just defined `to_Lf` to be the identity map, e.g., `lambda z: z` (say), then you could have the code `points = [...]` exactly once outside the `if/else`, rather than twice.   It seems like you've put a lot of work into making the if/else nicely symmetrical in both cases, so might as well go all the way.